# <center> Assignment 



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 

import warnings 
warnings.filterwarnings("ignore")
%matplotlib inline

In [2]:
df = pd.read_csv("Bengaluru_House_Data.csv")
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [3]:
df.shape

(13320, 9)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [5]:
# Check for missing values
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [6]:
df.corr()

bath   balcony     price
bath     1.000000  0.204201  0.456345
balcony  0.204201  1.000000  0.120355
price    0.456345  0.120355  1.000000

In [7]:
df.describe()

bath       balcony         price
count  13247.000000  12711.000000  13320.000000
mean       2.692610      1.584376    112.565627
std        1.341458      0.817263    148.971674
min        1.000000      0.000000      8.000000
25%        2.000000      1.000000     50.000000
50%        2.000000      2.000000     72.000000
75%        3.000000      2.000000    120.000000
max       40.000000      3.000000   3600.000000

# Insights:
- The average price for the properties is 112.57, with a standard deviation of 148.97. The minimum price is 8, and the maximum price is 3600.

-----

# <center>Answers of all Questions

# Answer 1
- In the case of SVM regression model, the most appropriate regression metric to employ would be the Mean Squared Error (MSE) or Root Mean Squared Error (RMSE).


- The reason for this is that SVM regression aims to minimize the distance between the predicted values and the actual values of the target variable. The MSE or RMSE measures the average squared (or square root of squared) difference between the predicted and actual values, which is a good indicator of the overall accuracy of the model.


- Other possible regression metrics that could be employed in this scenario include Mean Absolute Error (MAE), Mean Absolute Percentage Error (MAPE), and R-squared. However, MSE and RMSE are generally preferred over these other metrics for SVM regression models because they are sensitive to the magnitude of errors and help to penalize larger errors more heavily.






----------

# Answer 2
MSE,

If the goal is to predict the actual price of a house as accurately as possible, then the Mean Squared Error (MSE) would be a more appropriate evaluation metric to use for the SVM regression model.

- MSE measures the average squared difference between the predicted and actual values, and is directly related to the accuracy of the model's predictions. The lower the MSE value, the better the model's predictions are in terms of their closeness to the actual values.

- On the other hand, R-squared measures the proportion of variance in the target variable that is explained by the model. While R-squared is a useful metric for evaluating the overall goodness of fit of a regression model, it may not be as useful in evaluating the accuracy of individual predictions.

- Therefore, if the primary goal is to predict the actual price of a house as accurately as possible, then the MSE metric should be used to evaluate the performance of the SVM regression model.


------

# Answer 3
When dealing with a dataset that contains a significant number of outliers, Mean Absolute Error (MAE) would be the most appropriate regression metric to use with an SVM model.

- MAE is less sensitive to outliers compared to other regression metrics such as Mean Squared Error (MSE) or Root Mean Squared Error (RMSE). This is because MAE measures the average absolute difference between the predicted and actual values, rather than the squared difference as in MSE or RMSE.


- Outliers can have a significant impact on the squared difference, as they have a disproportionately large effect on the overall error calculation. By using MAE instead, the impact of outliers on the error calculation is reduced, resulting in a more robust evaluation of the model's performance.

- Therefore, when dealing with a dataset that contains a significant number of outliers, MAE is a better choice for evaluating the performance of an SVM regression model.

-------

# Answer 4
- If the values of both Mean Squared Error (MSE) and Root Mean Squared Error (RMSE) are very close, it is generally recommended to use RMSE as the evaluation metric for the SVM regression model with a polynomial kernel.

- The reason for this is that RMSE has the advantage of being expressed in the same units as the target variable, while MSE is expressed in squared units. This means that the RMSE value is easier to interpret, as it represents the average magnitude of the errors in the same units as the target variable



-----

# Answer 5 

If the goal is to measure how well the SVM regression model explains the variance in the target variable, the most appropriate evaluation metric would be the coefficient of determination, commonly known as R-squared.

- R-squared measures the proportion of variance in the target variable that is explained by the model. A high R-squared value indicates that the model is able to explain a larger proportion of the variance in the target variable, which means that it is a better fit for the data.

- In the case of comparing the performance of different SVM regression models with different kernels (linear, polynomial, and RBF), R-squared would be the most appropriate metric to use to evaluate the goodness of fit of each model. This will provide a measure of how well each model is able to explain the variance in the target variable, allowing for a comparison of their relative performance.